Sources/ReleaseServer/ReleaseServer.WebApi/Services/FsReleaseArtifactService.cs

```diff
@@ -152,23 +155,80 @@ public async Task<BackupInformationModel> RunBackup()
 
         public async Task RestoreBackup(IFormFile payload)
         {
-            using (var zipMapper = new ZipArchiveMapper())
+            
+            Logger.LogDebug("convert the uploaded payload to a ZIP archive");
+            using var fileStream = payload.OpenReadStream();
+            using var zipPayload = new ZipArchive(payload.OpenReadStream());
+            
+            await DirectoryLock.WaitAsync();
+
+            //It's important to release the semaphore. try / finally block ensures a guaranteed release (also if the operation may crash) 
+            try
             {
-                Logger.LogDebug("convert the uploaded backup payload to a ZIP archive");
-                var zipPayload = zipMapper.FormFileToZipArchive(payload);
+                await Task.Run(() => FsReleaseArtifactRepository.RestoreBackup(zipPayload));
+            }
+            finally
+            {
+                DirectoryLock.Release();
+            }
+        
+        }
 
-                await DirectoryLock.WaitAsync();
+        public async Task<ValidationResult> ValidateUploadPayload(IFormFile payload)
+        {
+            DeploymentMetaInfo deploymentMetaInfo;
+            
+            Logger.LogDebug("convert the uploaded payload to a ZIP archive");
+            using var fileStream = payload.OpenReadStream();
+            using var payloadZipArchive = new ZipArchive(payload.OpenReadStream());
+            
+            //Try to get the deployment.json entry from the zip archive and check, if it exists
+            var deploymentInfoEntry = payloadZipArchive.GetEntry("deployment.json");
+
+            if (deploymentInfoEntry == null)
+            {
+                var validationError = "the deployment.json does not exist in the uploaded payload!";
+                Logger.LogError(validationError);
+                return new ValidationResult {IsValid = false, ValidationError = validationError};
+            }
 
-                //It's important to release the semaphore. try / finally block ensures a guaranteed release (also if the operation may crash) 
+            //Open the deployment.json and extract the DeploymentMetaInfo of it
+            using (StreamReader deploymentInfoFile = new StreamReader(deploymentInfoEntry.Open(), System.Text.Encoding.UTF8))
+            {
                 try
                 {
-                    await Task.Run(() => FsReleaseArtifactRepository.RestoreBackup(zipPayload));
+                    JsonSerializer serializer = new JsonSerializer();
+                    deploymentMetaInfo = await Task.Run(() =>
+                        (DeploymentMetaInfo) serializer.Deserialize(deploymentInfoFile, typeof(DeploymentMetaInfo)));
```

Comment:
It does not make sense to schedule a new task for this. Please make this synchronous.

Reply:
Okay, i will do this

Reply:
ValidateUploadPayload() is now synchronous, see https://github.com/Traeger-GmbH/release-server/pull/58/commits/f79a7b0384bf265984b2328cf1923beea7e9d023